File: _exercises/23_quiz-game-5/23_quiz-game-5-app.R
library(shiny)
library(bslib)
library(ellmer)
library(shinychat)

# UI ---------------------------------------------------------------------------

ui <- page_sidebar(
  title = "Quiz Game 5",
  sidebar = sidebar(
    position = "right",
    fillable = TRUE,
    width = 400,
    value_box(
      "Correct Answers",
      textOutput("txt_correct"),
      showcase = fontawesome::fa_i("circle-check"),
      theme = "success"
    ),
    value_box(
      "Incorrect Answers",
      textOutput("txt_incorrect"),
      showcase = fontawesome::fa_i("circle-xmark"),
      theme = "danger"
    )
  ),
  navset_card_tab(
    nav_panel("Quiz Game", chat_mod_ui("chat")),
    nav_panel("Your Answers", tableOutput("tbl_scores"))
  )
)

# Server -----------------------------------------------------------------------

server <- function(input, output, session) {
  client <- chat(
    "anthropic/claude-3-7-sonnet-20250219",
    system_prompt = interpolate_file(
      # Use your quiz game system prompt, or switch to `_solutions` to use ours
      here::here("_exercises/14_quiz-game-1/prompt.md")
    ) |>
      # STEP 1: Add instructions about when to use the score-keeping tool ----
      paste(
        "\n\nAfter every question, use the 'Update Score' tool to... ____"
      )
  )

  scores <- reactiveVal(
    data.frame(
      theme = character(),
      question = character(),
      answer = character(),
      your_answer = character(),
      is_correct = logical()
    )
  )

  output$tbl_scores <- renderTable(scores())
  output$txt_correct <- renderText(sum(scores()$is_correct, na.rm = TRUE))
  output$txt_incorrect <- renderText(sum(!scores()$is_correct, na.rm = TRUE))

  # STEP 2: Implement the score-keeping tool ----
  # When the tool is called, add a new row to the `scores()` reactive data frame
  update_score <- function(theme, question, answer, your_answer, is_correct) {
    the_scores <- isolate(scores())
    the_scores <- rbind(
      the_scores,
      ____
    )
    # Now that we have new scores, update the `scores()` reactive value
    scores(the_scores)
    correct <- sum(the_scores$answer == the_scores$your_answer)
    # And return the current tally of correct and incorrect answers
    list(correct = correct, incorrect = nrow(the_scores) - correct)
  }

  client$register_tool(tool(
    update_score,
    description = paste(
      "Add a correct or incorrect answer to the score tally.",
      "Call this tool after you've graded the user's answer to a question."
    ),
    # STEP 3: Complete the tool definition ----
    arguments = list(
      theme = ____("The theme of the round."),
      question = ____("The quiz question that was asked."),
      answer = ____("The correct answer to the question."),
      your_answer = ____("The user's answer to the question."),
      is_correct = ____("Whether the user's answer was correct.")
    ) #,
    ## STEP 4: Add tool annotations ----
    # annotations = tool_annotations(
    #   title = "____",
    #   # https://fontawesome.com/search?q=plus&ic=free&o=r
    #   ____ = fontawesome::fa_i("____")
    # )
  ))

  chat <- chat_mod_server("chat", client)

  observe({
    # Note: This block starts the game when the app launches
    chat$update_user_input(
      value = "Let's play the quiz game!",
      submit = TRUE
    )
  })
}

shinyApp(ui, server)

File: _exercises/23_quiz-game-5/23_quiz-game-5-app.py
from typing import TypedDict

import chatlas
import dotenv
import polars as pl
from faicons import icon_svg
from pyhere import here
from shiny import App, reactive, render, ui

dotenv.load_dotenv()

# UI ---------------------------------------------------------------------------

app_ui = ui.page_sidebar(
    ui.sidebar(
        ui.value_box(
            "Correct Answers",
            ui.output_text("txt_correct"),
            showcase=icon_svg("circle-check"),
            theme="success",
        ),
        ui.value_box(
            "Incorrect Answers",
            ui.output_text("txt_incorrect"),
            showcase=icon_svg("circle-xmark"),
            theme="danger",
        ),
        position="right",
        fillable=True,
        width=400,
    ),
    ui.navset_card_underline(
        ui.nav_panel(
            "Quiz Game",
            ui.chat_ui("chat"),
        ),
        ui.nav_panel(
            "Your Answers",
            ui.output_data_frame("tbl_score"),
        ),
    ),
    title="Quiz Game 5",
    fillable=True,
)


class QuestionAnswer(TypedDict):
    theme: str
    question: str
    answer: str
    your_answer: str
    is_correct: bool


def server(input, output, session):
    chat_ui = ui.Chat(id="chat")

    # Set up the chat instance
    client = chatlas.ChatAnthropic(
        model="claude-3-7-sonnet-20250219",
        # Use your quiz game system prompt, or switch to _solutions to use ours
        # STEP 1: Add instructions about when to use the score-keeping tool ----
        system_prompt=f"""
{here("_exercises/14_quiz-game-1/prompt.md").read_text()}

After every question, use the "Update Score" tool to... ____
""",
    )

    scores = reactive.value[list[QuestionAnswer]]([])

    @render.data_frame
    def tbl_score():
        df = pl.DataFrame(scores())
        return df

    @render.text
    def txt_correct() -> int:
        return len([d for d in scores() if d["is_correct"]])

    @render.text
    def txt_incorrect() -> int:
        return len([d for d in scores() if not d["is_correct"]])

    # STEP 3: Complete the tool definition by filling in the type hints ----
    def update_score(
        theme: ____,
        question: ____,
        answer: ____,
        your_answer: ____,
        is_correct: ____,
    ):
        """
        Add a correct or incorrect answer to the score. Call this tool after
        you've graded the user's answer to a question.

        Parameters
        ----------
        theme: The theme of the round.
        question: The quiz question that was asked.
        answer: The correct answer to the question.
        your_answer: The user's answer to the question.
        is_correct: Whether the user's answer was correct.
        """
        with reactive.isolate():
            val_scores = scores.get()

        # STEP 2: Implement the score-keeping tool ----
        # When the tool is called, add a new answer to the scores reactive.value
        answer = QuestionAnswer(____)

        val_scores = [*val_scores, ____]
        scores.set(val_scores)

        correct = len([d for d in val_scores if d["is_correct"]])
        incorrect = len(val_scores) - correct
        return {"correct": correct, "incorrect": incorrect}

    client.register_tool(
        update_score,
        ## STEP 4: Add tool annotations ----
        # annotations={
        #     "title": "____",
        #     # https://fontawesome.com/search?q=plus&ic=free&o=r
        #     "extra": {"____": icon_svg("____")},
        # },
    )

    @chat_ui.on_user_submit
    async def handle_user_input(user_input: str):
        response = await client.stream_async(user_input, content="all")
        await chat_ui.append_message_stream(response)

    @reactive.effect
    def _():
        chat_ui.update_user_input(value="Let's play the quiz game!", submit=True)


app = App(app_ui, server)
